## Predicting Customer Churn at SyrialTel Using Machine Learning.
## 1. Business Understanding

Customer churn is a major challenge for SyriaTel, a telecommunications company, as losing customers directly impacts revenue and growth. Retaining existing customers is often more cost-effective than acquiring new ones.

SyriaTel faces the challenge of not being able to effectively identify customers who are at risk of leaving before they churn. This limits the company’s ability to take timely and proactive action to retain valuable customers.

This project aims to help SyriaTel identify customers who are likely to churn using machine learning techniques. By predicting churn in advance, the company can implement targeted marketing strategies, improve customer service, and offer personalized retention solutions.

A key focus of this project is minimizing missed churners, as failing to identify these customers represents lost revenue and missed retention opportunities.

### 1.1 Stakeholder
The primary stakeholders are SyriaTel’s customer retention and marketing teams, who are responsible for reducing customer churn and improving customer satisfaction.

### 1.2 Business Problem

SyriaTel is experiencing customer churn, leading to revenue loss and increased costs associated with acquiring new customers. The core challenge is the inability to accurately identify customers who are likely to leave in advance, limiting the company’s ability to intervene effectively.

### 1.3 Business Objective

The objective of this project is to develop a predictive model that accurately identifies customers at risk of churning, with a strong focus on minimizing missed churners. This will enable SyriaTel to take proactive retention actions, such as targeted marketing, improved customer support, and personalized offers, to reduce churn and protect revenue.

## 2. Data Understanding

### 2.1 Data Source
The dataset used in this project contains customer information from SyriaTel, a telecommunications company. It includes records of customer behavior, service usage, and account characteristics.


### 2.1 Loading Dataset
The dataset was loaded and examined to assess its structure, quality, and suitability for modeling.

In [1]:
import pandas as pd

# Load dataset
df = pd.read_csv('bigml_59c28831336c6604c800002a.csv')

# Preview first rows
df.head()

,state,account length,area code,phone number,international plan,voice mail plan,number vmail messages,total day minutes,total day calls,total day charge,...,total eve calls,total eve charge,total night minutes,total night calls,total night charge,total intl minutes,total intl calls,total intl charge,customer service calls,churn
0,KS,128,415,382-4657,no,yes,25,265.1,110,45.07,...,99,16.78,244.7,91,11.01,10.0,3,2.70,1,False
1,OH,107,415,371-7191,no,yes,26,161.6,123,27.47,...,103,16.62,254.4,103,11.45,13.7,3,3.70,1,False
2,NJ,137,415,358-1921,no,no,0,243.4,114,41.38,...,110,10.30,162.6,104,7.32,12.2,5,3.29,0,False
3,OH,84,408,375-9999,yes,no,0,299.4,71,50.90,...,88,5.26,196.9,89,8.86,6.6,7,1.78,2,False
4,OK,75,415,330-6626,yes,no,0,166.7,113,28.34,...,122,12.61,186.9,121,8.41,10.1,3,2.73,3,False


The dataset contains customer-level information for SyriaTel, including usage patterns, account details, and customer service interactions.

The target variable is 'churn', which indicates whether a customer has left the service.

In [2]:
#Initial Data Inspection
df.shape
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3333 non-null   object 
 1   account length          3333 non-null   int64  
 2   area code               3333 non-null   int64  
 3   phone number            3333 non-null   object 
 4   international plan      3333 non-null   object 
 5   voice mail plan         3333 non-null   object 
 6   number vmail messages   3333 non-null   int64  
 7   total day minutes       3333 non-null   float64
 8   total day calls         3333 non-null   int64  
 9   total day charge        3333 non-null   float64
 10  total eve minutes       3333 non-null   float64
 11  total eve calls         3333 non-null   int64  
 12  total eve charge        3333 non-null   float64
 13  total night minutes     3333 non-null   float64
 14  total night calls       3333 non-null   

In [3]:
# Checking for Missing Values
df.isnull().sum()

state                     0
account length            0
area code                 0
phone number              0
international plan        0
voice mail plan           0
number vmail messages     0
total day minutes         0
total day calls           0
total day charge          0
total eve minutes         0
total eve calls           0
total eve charge          0
total night minutes       0
total night calls         0
total night charge        0
total intl minutes        0
total intl calls          0
total intl charge         0
customer service calls    0
churn                     0
dtype: int64

In [4]:
# Checking for duplicates
df.duplicated().sum()

0

In [5]:
#Summary statistics
df.describe()

,account length,area code,number vmail messages,total day minutes,total day calls,total day charge,total eve minutes,total eve calls,total eve charge,total night minutes,total night calls,total night charge,total intl minutes,total intl calls,total intl charge,customer service calls
count,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000
mean,101.064806,437.182418,8.099010,179.775098,100.435644,30.562307,200.980348,100.114311,17.083540,200.872037,100.107711,9.039325,10.237294,4.479448,2.764581,1.562856
std,39.822106,42.371290,13.688365,54.467389,20.069084,9.259435,50.713844,19.922625,4.310668,50.573847,19.568609,2.275873,2.791840,2.461214,0.753773,1.315491
min,1.000000,408.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,23.200000,33.000000,1.040000,0.000000,0.000000,0.000000,0.000000
25%,74.000000,408.000000,0.000000,143.700000,87.000000,24.430000,166.600000,87.000000,14.160000,167.000000,87.000000,7.520000,8.500000,3.000000,2.300000,1.000000
50%,101.000000,415.000000,0.000000,179.400000,101.000000,30.500000,201.400000,100.000000,17.120000,201.200000,100.000000,9.050000,10.300000,4.000000,2.780000,1.000000
75%,127.000000,510.000000,20.000000,216.400000,114.000000,36.790000,235.300000,114.000000,20.000000,235.300000,113.000000,10.590000,12.100000,6.000000,3.270000,2.000000
max,243.000000,510.000000,51.000000,350.800000,165.000000,59.640000,363.700000,170.000000,30.910000,395.000000,175.000000,17.770000,20.000000,20.000000,5.400000,9.000000


In [6]:
#Categorical Variables
df.select_dtypes(include='object').nunique()
df.select_dtypes(include='object').head()

,state,phone number,international plan,voice mail plan
0,KS,382-4657,no,yes
1,OH,371-7191,no,yes
2,NJ,358-1921,no,no
3,OH,375-9999,yes,no
4,OK,330-6626,yes,no


In [7]:
# Target Variable (Churn) Distribution
df['churn'].value_counts()
df['churn'].value_counts(normalize=True)

False    0.855086
True     0.144914
Name: churn, dtype: float64

- The dataset contains 3,333 observations and 21 features.
- No missing values were found across any columns.
- No duplicate records were identified.
- The target variable, 'churn' is a binary variable indicating whether a customer has left the service.

Categorical variables such as 'international plan' and 'voice mail plan' are represented as "yes"/"no" and will require encoding for modeling.

The dataset includes a mix of customer usage metrics, account information, and service interaction data, which are all relevant for predicting customer churn.

 ### 2.2 Relevance to Business Problem

The features in this dataset are directly relevant to understanding customer behavior and potential dissatisfaction. For example:

- High customer service call frequency may indicate unresolved issues
- Usage patterns (minutes and charges) may reflect engagement levels
- Subscription plans (e.g., international or voicemail plans) may influence customer satisfaction

These variables provide meaningful signals that can help predict whether a customer is likely to churn, making the dataset well-suited for addressing the business problem of customer retention.

## 3. Data Preparation

In this section, the dataset is cleaned and transformed to prepare it for machine learning modeling. This includes handling irrelevant features, encoding categorical variables, and splitting the data into training and testing sets.

### 3.1 Dropping Irrelevant Columns
Dropping irrelevant colmuns suct as 'State' and 'Phone numbers' that do not contribute to churn prediction in this project.

In [8]:
# Dropping Columns
df = df.drop(['phone number', 'state'], axis=1)

### 3.2 Encoding Categorical Variables
Encoding categorical variables into numerical format for compatibility with machine learning algorithms.

In [9]:
# Encoding Categorical Variables
df['international plan'] = df['international plan'].map({'yes': 1, 'no': 0})
df['voice mail plan'] = df['voice mail plan'].map({'yes': 1, 'no': 0})

### 3.3 Defining Features (X) and target (y)

In [10]:
X = df.drop('churn', axis=1)
y = df['churn']

### 3.4 Train-Test Split
The dataset was split into training and testing sets to evaluate model performance on unseen data.

In [11]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.2, random_state=42)

### 3.5 Scaling for Logistic Regression
Feature scaling was applied to ensure that all variables contribute equally to the model. Caution was taken to avoid data leakage by fitting transformations only on the training data and applying them to the test data.

In [12]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [13]:
X_train.shape, X_test.shape

((2666, 18), (667, 18))

### 4.Modeling

### 4.1 Baseline Model: Logistic Regression

Selected Logistic Regression as the baseline model because it is simple, interpretable, and appropriate for binary classification. This model provides a starting point for comparison with more complex models.

Before fitting Logistic Regression, the numeric features were scaled because this model can be sensitive to differences in feature magnitude.

### 4.1.1 Logistic Regression Model Training

The model is trained using the training data only. This allows the model to learn patterns from known customer behavior without seeing the test data.

Keeping the test data separate helps evaluate how well the model may perform on new, unseen customers.

In [14]:
from sklearn.linear_model import LogisticRegression

# Initialize model
log_model = LogisticRegression(max_iter=1000)

# Train model
log_model.fit(X_train_scaled, y_train)

LogisticRegression(max_iter=1000)

### 4.1.2 Prediction

After training, the model is used to predict churn outcomes for the test set. These predictions are then compared to the actual churn values to measure model performance.

In [15]:
y_pred = log_model.predict(X_test_scaled)
y_pred

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False,  True, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False,  True, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False,  True,

### 4.1.3 Evaluating the Model

The model is evaluated using classification metrics, including accuracy, precision, recall, and F1-score.

For this business problem, recall is especially important because it measures how well the model identifies customers who actually churn. A false negative means the model failed to identify a customer who was likely to leave, which could result in lost revenue for SyriaTel.

Precision is also important because it shows how many predicted churners were actually churners. However, in this context, missing a churner is likely more costly than mistakenly flagging a customer for retention outreach.

In [16]:
from sklearn.metrics import classification_report, confusion_matrix

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

       False       0.87      0.98      0.92       566
        True       0.60      0.18      0.27       101

    accuracy                           0.86       667
   macro avg       0.73      0.58      0.60       667
weighted avg       0.83      0.86      0.82       667



### 4.1.4 Logistic Regression Model Performance
The Logistic Regression model achieved an overall accuracy of 86%. However, accuracy alone is not a reliable metric in this case due to class imbalance, as there are significantly more non-churners (566) than churners (101).

Focusing on the churn class (True):

- Precision: 0.60  
- Recall: 0.18  
- F1-score: 0.27  

The precision of 0.60 indicates that when the model predicts a customer will churn, it is correct 60% of the time. However, the recall of 0.18 shows that the model only identifies 18% of actual churners, meaning that 82% of customers who churn are not detected.

This is a significant limitation from a business perspective, as missing churners represents lost opportunities for intervention and customer retention.

The low F1-score further reflects the model’s poor balance between precision and recall in identifying churners.

Overall, while the model performs well at identifying non-churners, it is not effective at detecting customers at risk of leaving. Therefore, an improved model is needed to increase recall and better support customer retention efforts.

### 4.2 Improved Model: Decision Tree Classifier

After building the Logistic Regression baseline model, I created a Decision Tree model to test whether a non-linear model could better identify customers likely to churn.

A Decision Tree is useful because it can capture more complex relationships between customer behavior and churn. It is also relatively interpretable, which is helpful for explaining results to business stakeholders.

### 4.2.1 Training the Decision Tree Model
The Decision Tree model was trained using the training data only. Unlike Logistic Regression, Decision Trees do not require feature scaling because they split data based on feature thresholds rather than distance or coefficient size.

In [17]:
from sklearn.tree import DecisionTreeClassifier

# Initialize decision tree model
tree_model = DecisionTreeClassifier(random_state=42)

# Train model on unscaled training data
tree_model.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

### 4.2.2 Predictions
After training, the Decision Tree model was used to predict churn outcomes for customers in the test set. These predictions were compared to the actual churn values to evaluate model performance.

In [18]:
# predictions on the test data
tree_preds = tree_model.predict(X_test)
tree_preds

array([ True, False,  True, False, False, False, False, False, False,
       False, False, False, False,  True, False, False, False,  True,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False,  True, False, False, False,
       False, False,  True, False, False, False, False, False, False,
       False,  True,  True, False, False, False,  True, False, False,
       False, False, False, False, False, False,  True, False, False,
       False, False, False, False, False,  True,  True, False, False,
       False,  True, False, False, False, False,  True, False, False,
       False, False, False, False, False, False, False, False,  True,
       False,  True,  True, False, False,  True, False, False,  True,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False,  True, False, False, False, False, False, False, False,
       False,  True,

### 4.2.3 Evaluating the Decision Tree
The Decision Tree model was evaluated using the same classification metrics as the baseline model. This allows a fair comparison between models.

Recall remains especially important because SyriaTel wants to identify as many customers likely to churn as possible before they leave.

In [19]:
from sklearn.metrics import classification_report, confusion_matrix

print(classification_report(y_test, tree_preds))
confusion_matrix(y_test, tree_preds)

              precision    recall  f1-score   support

       False       0.95      0.95      0.95       566
        True       0.73      0.73      0.73       101

    accuracy                           0.92       667
   macro avg       0.84      0.84      0.84       667
weighted avg       0.92      0.92      0.92       667



array([[539,  27],
       [ 27,  74]], dtype=int64)

### 4.2.4 Decision Tree Model Performance

The Decision Tree model shows a clear improvement over the Logistic Regression baseline, achieving an overall accuracy of 92%.

Focusing on the churn class (True):

- Precision: 0.73  
- Recall: 0.73  
- F1-score: 0.73  

The recall of 0.73 indicates that the model correctly identifies 73% of customers who churn, which is a substantial improvement compared to the 18% recall from the Logistic Regression model. This means far fewer churners are missed.

Precision also improved to 0.73, meaning that when the model predicts a customer will churn, it is correct 73% of the time. This reflects a better balance between identifying churners and avoiding false alarms.

The F1-score of 0.73 further confirms that the model achieves a strong balance between precision and recall.

From a business perspective, this improvement is significant. The model is now able to identify a much larger proportion of at-risk customers, allowing SyriaTel to take proactive retention actions and reduce potential revenue loss.

Overall, the Decision Tree model performs substantially better than the baseline model and provides more practical value for identifying customers likely to churn.

### 4.3 Decision Tree Hyperparameter Tuning

The initial Decision Tree may overfit the training data because it can keep splitting until it captures very specific patterns. To reduce overfitting, I tuned hyperparameters that control tree complexity.


In [20]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'max_depth': [3, 5, 7, 10, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 5, 10]
}

grid_tree = GridSearchCV(
    estimator=DecisionTreeClassifier(random_state=42),
    param_grid=param_grid,
    scoring='recall',
    cv=5
)

grid_tree.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=42),
             param_grid={'max_depth': [3, 5, 7, 10, None],
                         'min_samples_leaf': [1, 2, 5, 10],
                         'min_samples_split': [2, 5, 10]},
             scoring='recall')

In [21]:
#Checking best parameters
grid_tree.best_params_

{'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 5}

### 4.3.1 Evaluating Tuned Model
The tuned Decision Tree was evaluated on the test set after selecting the best hyperparameters from the training data. This helps estimate how the model may perform on new customers.

The results of this model should be compared with the baseline Logistic Regression model to determine whether tuning improved churn detection.

In [22]:
best_tree = grid_tree.best_estimator_

tuned_tree_preds = best_tree.predict(X_test)

print(classification_report(y_test, tuned_tree_preds))
confusion_matrix(y_test, tuned_tree_preds)

              precision    recall  f1-score   support

       False       0.95      0.95      0.95       566
        True       0.73      0.71      0.72       101

    accuracy                           0.92       667
   macro avg       0.84      0.83      0.84       667
weighted avg       0.92      0.92      0.92       667



array([[539,  27],
       [ 29,  72]], dtype=int64)

### 4.3.2 Tuned Decision Tree Model Performance

The tuned Decision Tree model achieved an overall accuracy of 92%, similar to the untuned Decision Tree model.

Focusing on the churn class (True):

- Precision: 0.73  
- Recall: 0.71  
- F1-score: 0.72  

The recall of 0.71 indicates that the tuned model correctly identifies 71% of churners, which is slightly lower than the 73% achieved by the untuned Decision Tree model.

While the tuned model does not improve recall, it provides a more controlled model by limiting complexity through hyperparameter tuning. This can help reduce the risk of overfitting and improve the model’s ability to generalize to new data.

The tuned model performs similarly to the original Decision Tree, but does not provide a meaningful improvement in identifying churners.

### 5. Evaluation Comparison of 3 Models
The models were compared using accuracy, precision, recall, and F1-score. While accuracy shows overall correctness, recall is especially important for this churn project because missing a likely churner means SyriaTel may lose the chance to retain that customer.

The final model will be selected based on the best balance of recall and overall performance.

In [23]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

model_comparison = {
    'Model': ['Logistic Regression', 'Decision Tree', 'Tuned Decision Tree'],
    'Accuracy': [
        accuracy_score(y_test, y_pred),
        accuracy_score(y_test, tree_preds),
        accuracy_score(y_test, tuned_tree_preds)
    ],
    'Precision': [
        precision_score(y_test, y_pred),
        precision_score(y_test, tree_preds),
        precision_score(y_test, tuned_tree_preds)
    ],
    'Recall': [
        recall_score(y_test, y_pred),
        recall_score(y_test, tree_preds),
        recall_score(y_test, tuned_tree_preds)
    ],
    'F1 Score': [
        f1_score(y_test, y_pred),
        f1_score(y_test, tree_preds),
        f1_score(y_test, tuned_tree_preds)
    ]
}

comparison_df = pd.DataFrame(model_comparison)
comparison_df

,Model,Accuracy,Precision,Recall,F1 Score
0,Logistic Regression,0.857571,0.600000,0.178218,0.274809
1,Decision Tree,0.919040,0.732673,0.732673,0.732673
2,Tuned Decision Tree,0.916042,0.727273,0.712871,0.720000


### 6. Conclusion
Three models were evaluated to predict customer churn: Logistic Regression, Decision Tree, and a Tuned Decision Tree.

The Logistic Regression model achieved an accuracy of approximately 86%, but performed poorly in identifying churners, with a recall of only 0.18. This means the model missed over 80% of customers who actually churned, making it ineffective for the business problem despite its relatively high accuracy.

The Decision Tree model showed a significant improvement across all metrics. It achieved the highest recall of 0.73, meaning it correctly identified 73% of churners. Precision also improved to 0.73, indicating a strong balance between identifying churners and minimizing false positives. This model also had the highest overall accuracy at approximately 92%.

The Tuned Decision Tree model performed similarly to the original Decision Tree, with slightly lower recall (0.71) and marginally lower overall performance. While tuning helped control model complexity, it did not result in a meaningful improvement in identifying churners.

### 7. Business Recommendation
Based on the evaluation results, the Decision Tree model was selected as the final model because it provides the best balance between accuracy and, most importantly, recall.

For SyriaTel, recall is the most critical metric because it reflects the model’s ability to identify customers who are likely to churn. The Decision Tree model captures the highest proportion of at-risk customers, significantly reducing the number of missed churners compared to the baseline model.

This means SyriaTel can use the model to proactively target customers who are likely to leave, allowing the company to implement retention strategies such as personalized offers, improved customer support, or service adjustments.

Although the Logistic Regression model showed reasonable accuracy, it failed to detect most churners and therefore does not provide practical value for the business. The Tuned Decision Tree, while more controlled, did not outperform the original model in terms of recall.

In conclusion, the Decision Tree model is the most effective solution for addressing the business problem, as it enables SyriaTel to better identify at-risk customers and take action to reduce customer churn and potential revenue loss.